### PR TITLE
Connection: Split pairwise and cloud agent information

### PR DESCRIPTION
**Pull request by Patrik-Stas**
## Change 1
- Change in-memory representation of connection object - previously information about pw did/vk and cloud agent did/vk were stored in `AgentInfo` which was part of individual connection protocol state structures. This caused unnecessarily leaking connection routing implementation details into state machine layer. 

This PR splits `AgentInfo` into two structures - `PairwiseInfo` and `CloudAgentInfo` whereas `PairwiseInfo` is owned by Connection state machine and `CloudAgentInfo` is owned by VCX Connection object. 

Breaking changes:
Serialized format of Connection Inviter SM in `RequestedState`
Serialized format of Connection Inviter SM in `RespondedState`

## Change 2
- Connection object builds cloud agent upon receiving Connection request, not upon sending connection response

## Change 3
- Temporarily disabled alternative way of Inviter in Request/Responded state of searching for messages in its connection bootstrap cloud agent. It's relatively edge case scenario which was however not properly tested. Will be restored with future PR with appropriate testing.

### CODE REVIEW

**Comment by Patrik-Stas** [comment]
@mirgee please review, the tests are passing, CI failing on that random cargo dependency download issue

File: libvcx/src/aries/handlers/connection/invitee/states/complete.rs [hunk 1]
```diff
@@ -1,5 +1,5 @@
 use crate::error::VcxResult;
-use crate::aries::handlers::connection::agent_info::AgentInfo;
+use crate::aries::handlers::connection::pairwise_info::PairwiseInfo;
```
Comment (mirgee) [inline libvcx/src/aries/handlers/connection/invitee/states/complete.rs:2]: Not used

File: libvcx/src/aries/handlers/connection/connection.rs [hunk 2]
```diff
@@ -359,12 +485,18 @@ impl Connection {
      */
     pub fn connect(&mut self) -> VcxResult<()> {
         trace!("Connection::connect >>> source_id: {}", self.source_id());
+        let pairwise_info = self.pairwise_info();
+        let cloud_agent = CloudAgentInfo::create(&pairwise_info)?;
+        let routing_keys = cloud_agent.routing_keys()?;
+        let agency_endpoint = cloud_agent.service_endpoint()?;
+        self.cloud_agent_info = cloud_agent;
```
Comment (mirgee) [inline libvcx/src/aries/handlers/connection/connection.rs:492]: Did we agree to deprecate this? (We can do it separately of course...)

File: libvcx/Cargo.toml [hunk 3]
```diff
@@ -20,7 +20,6 @@ agency = [] # migrate these to v2 agency
 agency_pool_tests = []
 agency_v2 = []
 pool_legacy_agency_tests = [] # migrate the test using this
-aries = []
```
Comment (mirgee) [inline libvcx/Cargo.toml:23]: 👍 I was planning to clean up the tags as well

**Review by mirgee — CHANGES_REQUESTED** [review]

File: libvcx/src/aries/handlers/connection/connection.rs [hunk 4]
```diff
@@ -1,18 +1,22 @@
 use std::collections::HashMap;
 
+use crate::api::VcxStateType;
```
[the remark this answers is not in the document]
Comment (Patrik-Stas) [reply libvcx/src/aries/handlers/connection/connection.rs:3]: indeed

File: libvcx/src/aries/handlers/connection/connection.rs [hunk 5]
```diff
@@ -56,38 +60,48 @@ impl Connection {
...
         Ok(connection)
     }
 
+    pub fn from_parts(source_id: String, pairwise_info: PairwiseInfo, cloud_agent_info: CloudAgentInfo, state: SmConnectionState) -> Connection {
+        match state {
+            SmConnectionState::Inviter(state) => {
+                Connection {
+                    cloud_agent_info,
+                    connection_sm: SmConnection::Inviter(SmConnectionInviter::from(source_id, pairwise_info, state)),
+                    autohop_enabled: true
```
[the remark this answers is not in the document]
Comment (Patrik-Stas) [reply libvcx/src/aries/handlers/connection/connection.rs:92]: added as argument

File: libvcx/src/aries/handlers/connection/legacy_agent_info.rs [hunk 6]
```diff
@@ -0,0 +1,7 @@
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct LegacyAgentInfo {
+    pub pw_did: String,
+    pub pw_vk: String,
+    pub agent_did: String,
+    pub agent_vk: String,
+}
```
[the remark this answers is not in the document]
Comment (Patrik-Stas) [reply libvcx/src/aries/handlers/connection/legacy_agent_info.rs:7]: Definitely doable, but I wanted to minimize amount of breaking changes in this PR (though couldn't avoid some). The structure is right now only used to express format of serialized connection.

File: libvcx/src/aries/handlers/connection/connection.rs [hunk 7]
```diff
@@ -340,17 +364,119 @@ impl Connection {
...
             SmConnection::Inviter(sm_inviter) => {
-                SmConnection::Inviter(sm_inviter.clone().step(Some(message.clone()))?)
+                let (sm_inviter, new_cloud_agent_info, can_autohop) = match message {
+                    Some(message) => match message {
+                        A2AMessage::ConnectionRequest(request) => {
+                            let new_pairwise_info = PairwiseInfo::create()?;
+                            let new_cloud_agent = CloudAgentInfo::create(&new_pairwise_info)?;
+                            let new_routing_keys = new_cloud_agent.routing_keys()?;
+                            let new_service_endpoint = new_cloud_agent.service_endpoint()?;
+                            let sm_connection = sm_inviter.handle_connection_request(request, &new_pairwise_info, new_routing_keys, new_service_endpoint)?;
```
[the remark this answers is not in the document]
Comment (Patrik-Stas) [reply libvcx/src/aries/handlers/connection/connection.rs:381]: I am okay with that, but let's do that in separate PR

File: libvcx/src/aries/handlers/connection/connection.rs (hunk 2 above)
Comment (Patrik-Stas) [reply libvcx/src/aries/handlers/connection/connection.rs:492]: yes I agree, I'll do that separately

**Review by mirgee — CHANGES_REQUESTED** [review]

File: libvcx/src/aries/handlers/connection/connection.rs [hunk 8]
```diff
@@ -56,38 +60,48 @@ impl Connection {
...
-    pub fn create_with_invite(source_id: &str, invitation: Invitation, autohop: bool) -> VcxResult<Connection> {
+    pub fn create_with_invite(source_id: &str, invitation: Invitation, autohop_enabled: bool) -> VcxResult<Connection> {
         trace!("Connection::create_with_invite >>> source_id: {}", source_id);
-
+        let pairwise_info = PairwiseInfo::create()?;
         let mut connection = Connection {
-            connection_sm: SmConnection::Invitee(SmConnectionInvitee::new(source_id, autohop))
+            cloud_agent_info: CloudAgentInfo::default(),
+            connection_sm: SmConnection::Invitee(SmConnectionInvitee::new(source_id, pairwise_info)),
+            autohop_enabled
```
[the remark this answers is not in the document]
Comment (mirgee) [reply libvcx/src/aries/handlers/connection/connection.rs]: OK, let's leave it this way, as we agreed

File: libvcx/src/aries/handlers/connection/connection.rs (hunk 2 above)
Comment (mirgee) [reply libvcx/src/aries/handlers/connection/connection.rs:491]: OK

**Review by mirgee — APPROVED** [review]
